File: tool-designer/src/main/kotlin/com/infinum/designer/ui/views/magnifier/MagnifierView.java
package com.infinum.designer.ui.views.magnifier;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Point;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.Toast;

import androidx.core.content.ContextCompat;

import com.google.android.material.textview.MaterialTextView;
import com.infinum.designer.R;

public class MagnifierView extends FrameLayout {

    private MaterialTextView mColorValueTextView;

    private Drawable mMagnifyingLens;
    private Bitmap mPixels;
    private Paint mBitmapPaint;
    private Paint mGridPaint;
    private Paint mPixelOutlinePaint;

    private Rect mSourcePreviewRect;
    private Rect mDestinationPreviewRect;
    private RectF mTargetPixelOutline;

    private Point mInsets;
    private Path mPreviewClipPath;

    private int mCenterPixelColor;

    public MagnifierView(Context context) {
        this(context, null);
    }

    public MagnifierView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MagnifierView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public MagnifierView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        mMagnifyingLens = context.getDrawable(R.drawable.designer_loop_ring);

        mBitmapPaint = new Paint();
        mBitmapPaint.setAntiAlias(false);
        mBitmapPaint.setDither(true);
        mBitmapPaint.setFilterBitmap(false);

        DisplayMetrics dm = getResources().getDisplayMetrics();
        mGridPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mGridPaint.setColor(ContextCompat.getColor(context, R.color.designer_colorPickerMagnifierGridColor));
        mGridPaint.setAlpha(128);
        mGridPaint.setStyle(Paint.Style.STROKE);
        mGridPaint.setStrokeWidth(1f * dm.density);
        mGridPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.DARKEN));

        mPixelOutlinePaint = new Paint();
        mPixelOutlinePaint.setColor(ContextCompat.getColor(context, R.color.designer_colorPickerMagnifierGridColor));
        mPixelOutlinePaint.setStyle(Paint.Style.STROKE);
        mPixelOutlinePaint.setStrokeWidth(2f * dm.density);
        mPixelOutlinePaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.DARKEN));

        final Resources res = getResources();
        mInsets = new Point(res.getDimensionPixelSize(R.dimen.designer_magnified_image_horizontal_inset),
                res.getDimensionPixelSize(R.dimen.designer_magnified_image_vertical_inset));

        int previewSize = res.getDimensionPixelSize(R.dimen.designer_magnified_image_size);
        mDestinationPreviewRect = new Rect(mInsets.x, mInsets.y, mInsets.x + previewSize,
                mInsets.y + previewSize);
        mPreviewClipPath = new Path();
        mPreviewClipPath.addCircle(mDestinationPreviewRect.exactCenterX(),
                mDestinationPreviewRect.exactCenterY(), previewSize / 2f, Path.Direction.CCW);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        mColorValueTextView = findViewById(R.id.color_value);
        mColorValueTextView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                ClipboardManager cm = getContext().getSystemService(ClipboardManager.class);
                CharSequence text = mColorValueTextView.getText();
                ClipData primaryClip = cm.getPrimaryClip();
                if (primaryClip == null || primaryClip.getItemAt(0) == null ||
                        !text.equals(cm.getPrimaryClip().getItemAt(0).coerceToText(getContext()))) {
                    ClipData clip = ClipData.newPlainText("color", text);
                    cm.setPrimaryClip(clip);
                    Toast.makeText(v.getContext(),
                            R.string.designer_color_copied_to_clipboard,
                            Toast.LENGTH_SHORT)
                            .show();
                }
            }
        });
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mMagnifyingLens.setBounds(0, 0, w, h);
    }

    @Override
    protected void dispatchDraw(Canvas canvas) {
        mMagnifyingLens.draw(canvas);
        if (mPixels != null) {
            canvas.clipPath(mPreviewClipPath);
            canvas.drawBitmap(mPixels, mSourcePreviewRect, mDestinationPreviewRect, mBitmapPaint);
            drawGrid(canvas);
            canvas.drawRect(mTargetPixelOutline, mPixelOutlinePaint);
        }
        super.dispatchDraw(canvas);
    }

    public void setPixels(Bitmap pixels) {
        mPixels = pixels;
        mSourcePreviewRect = new Rect(0, 0, pixels.getWidth(), pixels.getHeight());
        mCenterPixelColor = pixels.getPixel(pixels.getWidth() / 2, pixels.getHeight() / 2);

        if (mTargetPixelOutline == null) {
            float pixelSize = (float) mDestinationPreviewRect.width() / pixels.getWidth();
            float x = (mPixels.getWidth() - 1) / 2f * pixelSize;
            float y = (mPixels.getHeight() - 1) / 2f * pixelSize;
            mTargetPixelOutline = new RectF(mDestinationPreviewRect.left + x,
                    mDestinationPreviewRect.top + y, mDestinationPreviewRect.left + x + pixelSize,
                    mDestinationPreviewRect.top + y + pixelSize);
        }

        if (mColorValueTextView != null) {
            mColorValueTextView.setText(String.format("#%06X", mCenterPixelColor & 0x00ffffff));
        }
        invalidate();
    }

    private void drawGrid(Canvas canvas) {
        final float stepSize = mDestinationPreviewRect.width() / (float) mSourcePreviewRect.width();

        for (float x = mDestinationPreviewRect.left + stepSize;
             x <= mDestinationPreviewRect.right; x += stepSize) {
            canvas.drawLine(x, mDestinationPreviewRect.top, x, mDestinationPreviewRect.bottom,
                    mGridPaint);
        }
        for (float y = mDestinationPreviewRect.top + stepSize;
             y <= mDestinationPreviewRect.bottom; y += stepSize) {
            canvas.drawLine(mDestinationPreviewRect.left, y, mDestinationPreviewRect.right, y,
                    mGridPaint);
        }
    }
}

File: tool-designer/src/main/kotlin/com/infinum/designer/ui/views/magnifier/MagnifierView.kt
package com.infinum.designer.ui.views.magnifier

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Point
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.widget.FrameLayout
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.infinum.designer.R
import com.infinum.designer.databinding.DesignerLayoutMagnifierBinding
import com.infinum.designer.extensions.getHexCode
import com.infinum.designer.extensions.half
import com.infinum.designer.extensions.orZero

class MagnifierView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private val clipboardManager =
        context.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager

    private lateinit var viewBinding: DesignerLayoutMagnifierBinding

    private var centerPixelColor = 0

    private val previewSize = resources.getDimensionPixelSize(R.dimen.designer_magnified_image_size)

    private var lensDrawable: Drawable? = null

    private var currentBitmap: Bitmap? = null

    private val bitmapPaint: Paint = Paint().apply {
        isAntiAlias = false
        isDither = true
        isFilterBitmap = false
    }

    private val gridPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = ContextCompat.getColor(
            context,
            R.color.designer_colorPickerMagnifierGridColor
        )
        alpha = 128
        style = Paint.Style.STROKE
        strokeWidth = 1f * resources.displayMetrics.density
        xfermode = PorterDuffXfermode(PorterDuff.Mode.DARKEN)
    }

    private val outlinePaint: Paint = Paint().apply {
        color = ContextCompat.getColor(
            context,
            R.color.designer_colorPickerMagnifierGridColor
        )
        style = Paint.Style.STROKE
        strokeWidth = 2f * resources.displayMetrics.density
        xfermode = PorterDuffXfermode(PorterDuff.Mode.DARKEN)
    }

    private val insets: Point = Point(
        resources.getDimensionPixelSize(R.dimen.designer_magnified_image_inset),
        resources.getDimensionPixelSize(R.dimen.designer_magnified_image_inset)
    )

    private val previewRect: Rect = Rect(
        insets.x,
        insets.y,
        insets.x + previewSize,
        insets.y + previewSize
    )

    private val previewClipPath: Path = Path().apply {
        addCircle(
            previewRect.exactCenterX(),
            previewRect.exactCenterY(),
            previewSize.half().toFloat(),
            Path.Direction.CCW
        )
    }

    private var pixelOutline: RectF? = null

    private var sourceRect: Rect? = null

    override fun onFinishInflate() {
        super.onFinishInflate()

        viewBinding = DesignerLayoutMagnifierBinding.bind(this)

        with(viewBinding) {
            lensDrawable = root.background

            colorValueView.setOnClickListener {
                clipboardManager?.setPrimaryClip(
                    ClipData.newPlainText("color", colorValueView.text)
                )
                Toast.makeText(
                    it.context,
                    R.string.designer_color_copied_to_clipboard,
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        lensDrawable?.setBounds(0, 0, w, h)
    }

    override fun dispatchDraw(canvas: Canvas) {
        lensDrawable?.draw(canvas)
        currentBitmap?.let {
            with(canvas) {
                clipPath(previewClipPath)
                currentBitmap?.let { drawBitmap(it, sourceRect, previewRect, bitmapPaint) }
                drawGrid(canvas)
                pixelOutline?.let { drawRect(it, outlinePaint) }
            }
        }
        super.dispatchDraw(canvas)
    }

    fun setPixels(pixelsBitmap: Bitmap) {
        if (pixelsBitmap.width == 1 || pixelsBitmap.height == 1) {
            return
        }

        this.currentBitmap = pixelsBitmap

        sourceRect = Rect(0, 0, pixelsBitmap.width, pixelsBitmap.height)

        centerPixelColor = pixelsBitmap.getPixel(
            pixelsBitmap.width.half().coerceAtMost(pixelsBitmap.width),
            pixelsBitmap.height.half().coerceAtMost(pixelsBitmap.height)
        )

        if (pixelOutline == null) {
            val pixelSize = previewRect.width().toFloat() / pixelsBitmap.width
            currentBitmap?.let {
                val x = (it.width - 1).half() * pixelSize
                val y = (it.height - 1).half() * pixelSize
                pixelOutline = RectF(
                    previewRect.left + x,
                    previewRect.top + y, previewRect.left + x + pixelSize,
                    previewRect.top + y + pixelSize
                )
            }
        }

        viewBinding.colorValueView.text = centerPixelColor.getHexCode()

        invalidate()
    }

    private fun drawGrid(canvas: Canvas) {
        val stepSize = previewRect.width() / sourceRect?.width().orZero().coerceAtLeast(1).toFloat()
        var x = previewRect.left + stepSize
        while (x <= previewRect.right) {
            canvas.drawLine(
                x,
                previewRect.top.toFloat(),
                x,
                previewRect.bottom.toFloat(),
                gridPaint
            )
            x += stepSize
        }
        var y = previewRect.top + stepSize
        while (y <= previewRect.bottom) {
            canvas.drawLine(
                previewRect.left.toFloat(),
                y,
                previewRect.right.toFloat(),
                y,
                gridPaint
            )
            y += stepSize
        }
    }
}